pgxpool/pool.go

```diff
@@ -102,6 +106,60 @@ type Pool struct {
 	closeChan chan struct{}
 }
 
+// AutoloadAfterConnect is suitable for assigning to the AfterConnect configuration setting.
+// It will automatically load the named types for each connection in an efficient manner,
+// performing a single query to the database backend. The underlying call to pgx.LoadTypes
+// is smart enough to also retrieve any related types required to support the definition of the
+// named types.
+// If reuseTypeMap is enabled, it is assumed that the OID mapping is stable across all database
+// backends in this pool, resulting in only needing to query when creating the initial connection;
+// subsequent connections will reuse the same OID type mapping.
+// Because it is not always possible for a client to know the database topology in the final usage
+// context, PGXPOOL_REUSE_TYPEMAP, when given a value of y or n, will take precedence over this argument.
+func AutoloadAfterConnect(typeNames []string, reuseTypeMap bool) func(context.Context, *pgx.Conn) error {
+	switch os.Getenv("PGXPOOL_REUSE_TYPEMAP") {
+	case "y":
+		reuseTypeMap = true
+	case "n":
+		reuseTypeMap = false
+	}
+	if reuseTypeMap {
+		mutex := new(sync.Mutex)
+		var types []*pgtype.Type
+		return func(ctx context.Context, conn *pgx.Conn) error {
+			if types != nil {
+				// avoid acquiring the mutex if the types are already available
+				conn.TypeMap().RegisterTypes(types)
+				return nil
+			}
```

Comment:
I'm pretty sure this is not safe. The read on line 130 could happen concurrently with the write on line 144. I don't think there is any way around some sort of synchronization.

But also, this only happens when creating a connection. Any performance gains would be insignificant compared to the complexity beyond doing the simplest locking.

Reply:
I think this is safe. Whatever is happening in the mutex, either this variable will be set to `nil` (if there is nothing yet holding the mutex, or if that goroutine hasn't yet updated the variable), or it will be set to its final value.
If it's `nil`, we then acquire the mutex and proceed safely. But there is no need to acquire the mutex if the value has already been set.
The reason this is safe is that the transition is only in one direction, from `nil` to its final value.
I can't find any documentation indicating that a goroutine will get a runtime error or similar; just that the value could be changed "underneath it", which in the situation is fine, as it would result in the next step being to acquire the mutex and re-verify the value.
The purpose of this check before acquiring the mutex is to cater for the common case when the types are already known, allowing new connections to instantly access the types without needing to wait for any other connections also in the process of running this code.

Here's a silly little gist showing how there is no problem reading a shared variable without a mutex, as long as you don't care about timing: https://gist.github.com/nicois/d4aabaa6fc1c0ed998a1e4cc9bd102f5#file-no-mutex-go

Reply:
Try running it with the race detector:

```
jack@glados ~/dev/gorun » go run -race main.go
==================
WARNING: DATA RACE
Read at 0x000100606028 by goroutine 10:
  main.reader()
      /Users/jack/dev/gorun/main.go:38 +0xc0
  main.main.gowrap2()
      /Users/jack/dev/gorun/main.go:54 +0x4c

Previous write at 0x000100606028 by goroutine 6:
  main.writer()
      /Users/jack/dev/gorun/main.go:21 +0x94
  main.main.gowrap1()
      /Users/jack/dev/gorun/main.go:51 +0x4c

Goroutine 10 (running) created at:
  main.main()
      /Users/jack/dev/gorun/main.go:54 +0x150

Goroutine 6 (running) created at:
  main.main()
      /Users/jack/dev/gorun/main.go:51 +0x12c
==================
Read 238100 values
Read 246484 values
Read 239879 values
Read 237489 values
Read 238356 values
Read 230082 values
Written 10056 values
Read 242796 values
Read 231918 values
Read 234697 values
Read 266612 values
nothing unexpected
Found 1 data race(s)
exit status 66
```